src/GafferSceneUI/SceneView.cpp

```diff
@@ -1335,77 +1335,21 @@ void SceneView::frame( const GafferScene::PathMatcher &filter, const Imath::V3f
 
 void SceneView::expandSelection( size_t depth )
 {
-	Context::Scope scopedContext( getContext() );
-
 	PathMatcher &selection = const_cast<GafferScene::PathMatcherData *>( m_sceneGadget->getSelection() )->writable();
-	PathMatcher &expanded = expandedPaths()->writable();
 
-	std::vector<string> toExpand;
-	selection.paths( toExpand );
+	PathMatcher newlyExpandedLeaves = ContextAlgo::expandDescendants( getContext(), selection, preprocessedInPlug<ScenePlug>(), depth - 1 );
 
+	selection.clear();
 	bool needUpdate = false;
-	for( std::vector<string>::const_iterator it = toExpand.begin(), eIt = toExpand.end(); it != eIt; ++it )
+	for( PathMatcher::Iterator it = newlyExpandedLeaves.begin(), eIt = newlyExpandedLeaves.end(); it != eIt; ++it )
```

Comment:
I'm going to push some fixes shortly, but I just tested (a) remove scope from `expandDescendants` and instead scope it from `expandSelection` and (b) stop clearing the selection.

With the code in a temporary state:

```
void SceneView::expandSelection( size_t depth )
{
	PathMatcher &selection = const_cast<GafferScene::PathMatcherData *>( m_sceneGadget->getSelection() )->writable();

	Context::Scope scope( getContext() );
	PathMatcher newlyExpandedLeaves = ContextAlgo::expandDescendants( getContext(), selection, preprocessedInPlug<ScenePlug>(), depth - 1 );

//	selection.clear();
	if( selection.addPaths( newlyExpandedLeaves ) )
	{
		ContextAlgo::setSelectedPaths( getContext(), m_sceneGadget->getSelection()->readable() );
	}
}
```

and I got this failure:

```
FAIL: testExpandSelection (GafferSceneUITest.SceneViewTest.SceneViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.33.2.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferSceneUITest/SceneViewTest.py", line 108, in testExpandSelection
    self.assertEqual( getSelection(), set( [ "/A/B", "/A/C" ] ) )
AssertionError: Items in the first set but not the second:
'/A'
----------------------------------------------------------------------
```

which is right up the top of `testExpandSelection` : 

```
setExpandedPaths( [ "/" ] )
self.assertEqual( getExpandedPaths(), set( [ "/" ] ) )

# expand 1 level from root

view.expandSelection()

self.assertEqual( getExpandedPaths(), set( [ "/", "/A" ] ) )
self.assertEqual( getSelection(), set( [ "/A/B", "/A/C" ] ) ) # this line fails
```

If I instead uncomment the `selection.clear()` or assign directly as you suggest, the test suite passes.

```
void SceneView::expandSelection( size_t depth )
{
	PathMatcher &selection = const_cast<GafferScene::PathMatcherData *>( m_sceneGadget->getSelection() )->writable();

	Context::Scope scope( getContext() );
	selection = ContextAlgo::expandDescendants( getContext(), selection, preprocessedInPlug<ScenePlug>(), depth - 1 );

	ContextAlgo::setSelectedPaths( getContext(), selection );
}
```

There could still be a bug in `expandDescendants` that was causing said failure, and the selection clearing is masking it, but I guess there is no test showing the "previously selected paths that are already leaves are meant to stay selected" behavior?